### PR TITLE
Re-introduce configurable key swallowing

### PR DESCRIPTION
**Pull request by jwijenbergh**
Continuation of #2925 and #3102

~Instead of grabbing keys sync, which has problems in Dmenu (as it grabs the keyboard). Let's use XSendEvent. Note that mouse buttons are still grabbed sync, but this does not interfere with dmenu. If someone thinks this is unreasonable I could also use XSendEvent here.~

In the discussion here, we have figured out that XSendEvent would not work for every client. In this pr I'm using XTest

### CODE REVIEW

**Comment by jwijenbergh** [comment]
Simulated key presses are now proper X events.

simulatekeypress

They still go through the qtile keybindings, so if an user wants to ignore the swallowing when simulating keypresses, they can make their own function that ungrabs and then grabs the binding again. Or just mark the keybind non-swallowing with `swallow=False`.

File: libqtile/backend/x11/core.py
Comment (jwijenbergh) [inline libqtile/backend/x11/core.py:614]: I've added this to shush mypy but it still says Window has no attribute window? Any reason why?

Comment (m-col) [reply libqtile/backend/x11/core.py:614]: `hasattr` here will evaluate to `True` even if `self.qtile.current_window.window` is `None`. Maybe you were meaning to use `getattr`? You could do `assert self.qtile.current_window.window`. Or alternatively `if self.qtile.current_window.window:`

File: libqtile/backend/x11/core.py
Comment (m-col) [inline libqtile/backend/x11/core.py:787]: Perhaps it would be better to use the real pointer coordinates? Otherwise windows will always receive the events in the top left corner.

File: test/test_swallow.py
Comment (m-col) [inline test/test_swallow.py:29]: Idea: you could use a global variable as a counter and increment it every time this function is called. That way we can confirm that the bound function is being called when expected.

File: libqtile/backend/x11/core.py
Comment (jwijenbergh) [reply libqtile/backend/x11/core.py:787]: It is but they're keypresses not mouse presses so does this make a difference?

File: test/test_swallow.py
Comment (jwijenbergh) [reply test/test_swallow.py:29]: Good idea!

File: test/test_swallow.py
Comment (m-col) [inline test/test_swallow.py:94]: I wonder if we could replace this with `cmd_simulate_buttonpress` (once the wayland implementation is added)

Comment (jwijenbergh) [reply test/test_swallow.py:94]: Probably indeed! The way I did it here is really hacky

File: libqtile/backend/x11/core.py
Comment (m-col) [reply libqtile/backend/x11/core.py:787]: Ah yeah, my mistake. I assume it doesn't make a difference, and perhaps the arguments are there because the signature might match a generic synthetic event signature or something.

File: libqtile/backend/x11/core.py
Comment (jwijenbergh) [reply libqtile/backend/x11/core.py:614]: Oh that makes sense, I will change this

File: test/test_swallow.py
Comment (m-col) [reply test/test_swallow.py:94]: Ha, we have lots of these hacky evals in the tests - they are very useful!

Comment (jwijenbergh) [reply test/test_swallow.py:94]: True but it's especially hacky because I used None for the self argument in lookup key :laughing:

File: libqtile/backend/x11/core.py
Comment (jwijenbergh) [reply libqtile/backend/x11/core.py:787]: Well x11 does actually give the pointer x and y on real events when I logged them, so your idea is not wrong

File: test/test_swallow.py
Comment (jwijenbergh) [reply test/test_swallow.py:94]: On second thought, the `cmd_simulate_keypress` function does not return the handled handled boolean which is needed here. So that will have to be added there

Comment (m-col) [reply test/test_swallow.py:94]: The test window gets closed if it receives a key event, so if it's swallow vs non-swallow that you need to check then that might be a useful way to detect if a key event was not swallowed.

Comment (jwijenbergh) [reply test/test_swallow.py:94]: TIL that is useful!

File: libqtile/backend/x11/core.py
Comment (jwijenbergh) [reply libqtile/backend/x11/core.py]: Still doesn't work :(, I wanted to do hasattr as window is not defined as a member in the backend window base

Comment (m-col) [reply libqtile/backend/x11/core.py]: Maybe `assert getattr(self.qtile.current_window, "window", False)`? Then it's saying that the attribute exists and is truthy

Comment (jwijenbergh) [reply libqtile/backend/x11/core.py]: Good idea!, trying...

Comment (jwijenbergh) [reply libqtile/backend/x11/core.py]: Still the same, maybe there is some ignore comment we can do?

Comment (m-col) [reply libqtile/backend/x11/core.py]: Maybe `assert isinstance(self.qtile.current_window, (window.Window, window.Static))`? mypy is likely inferring that the window is the base window class. Otherwise `# type: ignore` on the offending lines, presumably where `current_window.window` is accessed.

Comment (jwijenbergh) [reply libqtile/backend/x11/core.py]: `assert isinstance(self.qtile.current_window, (window.Window, window.Static))` worked, thanks a lot!

File: test/test_swallow.py
Comment (jwijenbergh) [reply test/test_swallow.py:29]: I couldn't get a global variable working :/ I don't think pytest likes that, I tried a bunch of things. But the global variable is not correctly updated, it stays at 0. Even though the function is executed

File: libqtile/backend/x11/core.py
Comment (jwijenbergh) [reply libqtile/backend/x11/core.py]: Implemented.

File: test/test_swallow.py
Comment (elParaguayo) [reply test/test_swallow.py:29]: Can you make a fixture and extend the scope of the fixture to `module` - your noop function could then increment a value within that fixture.

Feel free to ignore this as I'm coming at it with virtually no context!

Comment (m-col) [reply test/test_swallow.py:29]: Oh yeah nice, good idea

Comment (jwijenbergh) [reply test/test_swallow.py:29]: Thanks. I tried something with fixtures but I hardly understood them. I will have another look at it. I guess the same could be done for the .when() Match pr, but there I just used `toggle_floating` with a test window.

**Comment by m-col** [comment]
Unfortunately this isn't working for me as expected. With this config to call rofi:
```py
keys.extend([
    Key(["shift"], "f", lazy.spawn("rofi -show"), swallow=True),
    Key(["control"], "f", lazy.spawn("rofi -show"), swallow=False),
])
```
The keys are not being sent to the focused window (xterm) when I call the second binding.

**Comment by jwijenbergh** [comment]
> :

Hmmm and does it work without a binding that uses rofi? Isn't this the expected behaviour because rofi grabs the keyboard? Because when I try to e.g. switch groups in dmenu (have to check rofi) it won't allow me to.

**Comment by m-col** [comment]
Oh, I'm being an idiot, aren't I? We decided that `swallow=False` is unsupported when the binding triggers a keyboard grab, didn't we? Please could you add a note about that to the changelog entry and the docstrings in config.py? People are almost certainly going to be confused by this!

**Comment by jwijenbergh** [comment]
> Oh, I'm being an idiot, aren't I? We decided that `swallow=False` is unsupported when the binding triggers a keyboard grab, didn't we? Please could you add a note about that to the changelog entry and the docstrings in config.py? People are almost certainly going to be confused by this!

Now that I think about it, if it doesn't work for you then that is actually not correct. `swallow=False` with dmenu is only a problem with sync grabs, as we're still grabbing async in this pr and just sending an event this should work. Furthermore, as you're using `lazy.spawn`, rofi/dmenu grabbing the keyboard shouldn't even be a problem so not what sure is going on for you

For me all seems to work fine.

## Starting rofi with lazy.spawn
Key:
```python
Key([], "b", lazy.spawn("rofi -show"), swallow=False),
```
Gif:
swallowfalserofi

## CommandSet
This also works and forwards the key to the focused window *after* dmenu is closed, which makes sense.
```python
Key([], "b", lazy.run_extension(extension.CommandSet(
    commands={
        'a': 'xterm'
    },
foreground='#FF0000', selected_background='#FF0000')),
swallow=False),
```

swallowfalsecommandset


Can you do this keybinding with the focused window being "xev" ? To see if the event is correctly forwarded.

File: test/test_swallow.py
Comment (jwijenbergh) [reply test/test_swallow.py:29]: @elParaguayo do you have a simple example how can I implement this with fixtures? I can't seem to access it inside the lazy function. How would the lazy function and fixture itself look like?

**Comment by m-col** [comment]
Using the same keybinding as you, `Key([], "b", lazy.spawn("rofi -show"), swallow=False),`, these are the events printed by xev:

```
FocusOut event, serial 34, synthetic NO, window 0x1c00001,
    mode NotifyGrab, detail NotifyAncestor

KeyPress event, serial 34, synthetic YES, window 0x1a3,
    root 0x1a3, subw 0x1c00001, time 23816063, (1495,633), root:(1495,633),
    state 0x0, keycode 56 (keysym 0x62, b), same_screen YES,
    XLookupString gives 1 bytes: (62) "b"
    XmbLookupString gives 1 bytes: (62) "b"
    XFilterEvent returns: False

LeaveNotify event, serial 34, synthetic NO, window 0x1c00001,
    root 0x1a3, subw 0x0, time 23816163, (530,102), root:(1495,633),
    mode NotifyGrab, detail NotifyAncestor, same_screen YES,
    focus YES, state 0

FocusIn event, serial 34, synthetic NO, window 0x1c00001,
    mode NotifyUngrab, detail NotifyAncestor

KeymapNotify event, serial 34, synthetic NO, window 0x0,
    keys:  2   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   

FocusOut event, serial 34, synthetic NO, window 0x1c00001,
    mode NotifyGrab, detail NotifyAncestor

FocusIn event, serial 34, synthetic NO, window 0x1c00001,
    mode NotifyUngrab, detail NotifyAncestor

KeymapNotify event, serial 34, synthetic NO, window 0x0,
    keys:  2   0   0   0   0   0   0   0   4   0   0   0   0   0   0   0   
           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   

EnterNotify event, serial 34, synthetic NO, window 0x1c00001,
    root 0x1a3, subw 0x0, time 23817793, (530,102), root:(1495,633),
    mode NotifyUngrab, detail NotifyAncestor, same_screen YES,
    focus YES, state 0

KeymapNotify event, serial 34, synthetic NO, window 0x0,
    keys:  0   0   0   0   0   0   0   0   4   0   0   0   0   0   0   0   
           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   

KeyRelease event, serial 34, synthetic NO, window 0x1c00001,
    root 0x1a3, subw 0x0, time 23817872, (530,102), root:(1495,633),
    state 0x0, keycode 66 (keysym 0xff1b, Escape), same_screen YES,
    XLookupString gives 1 bytes: (1b) "
    XFilterEvent returns: False

<< The events below are me focussing back to the terminal that called xev >>

KeyPress event, serial 34, synthetic NO, window 0x1c00001,
    root 0x1a3, subw 0x0, time 23819378, (530,102), root:(1495,633),
    state 0x0, keycode 64 (keysym 0xffeb, Super_L), same_screen YES,
    XLookupString gives 0 bytes: 
    XmbLookupString gives 0 bytes: 
    XFilterEvent returns: False

FocusOut event, serial 34, synthetic NO, window 0x1c00001,
    mode NotifyGrab, detail NotifyAncestor

PropertyNotify event, serial 34, synthetic NO, window 0x1c00001,
    atom 0x19a (_NET_FRAME_EXTENTS), time 23819452, state PropertyNewValue

ConfigureNotify event, serial 34, synthetic YES, window 0x1c00001,
    event 0x1c00001, window 0x1c00001, (960,526), width 950, height 516,
    border_width 5, above 0x0, override NO

PropertyNotify event, serial 34, synthetic NO, window 0x1c00001,
    atom 0x198 (WM_STATE), time 23819453, state PropertyNewValue

FocusOut event, serial 34, synthetic NO, window 0x1c00001,
    mode NotifyWhileGrabbed, detail NotifyNonlinear

FocusOut event, serial 34, synthetic NO, window 0x1c00001,
    mode NotifyUngrab, detail NotifyPointer
```

The second event looks like a 'b' is being sent to xev, which made me think that I *may* have read that xterm does not support send-event! However, testing with urxvt, kitty, and firefox, none of these show a 'b'. Here is a clip of me testing with firefox, which I include because you can actually see when I press the 'b' key that the focus to the text input kinda flashes off-on-off before rofi appears... not sure if that is relevant:

https://user-images.githubusercontent.com/44536113/146844275-73ded2b7-ecba-44c6-959c-371ac224eb9e.mp4

What terminal are you testing with?

**Comment by jwijenbergh** [comment]
> Using the same keybinding as you, `Key([], "b", lazy.spawn("rofi -show"), swallow=False),`, these are the events printed by xev:
> 
> ```
> FocusOut event, serial 34, synthetic NO, window 0x1c00001,
>     mode NotifyGrab, detail NotifyAncestor
> 
> KeyPress event, serial 34, synthetic YES, window 0x1a3,
>     root 0x1a3, subw 0x1c00001, time 23816063, (1495,633), root:(1495,633),
>     state 0x0, keycode 56 (keysym 0x62, b), same_screen YES,
>     XLookupString gives 1 bytes: (62) "b"
>     XmbLookupString gives 1 bytes: (62) "b"
>     XFilterEvent returns: False
> 
> LeaveNotify event, serial 34, synthetic NO, window 0x1c00001,
>     root 0x1a3, subw 0x0, time 23816163, (530,102), root:(1495,633),
>     mode NotifyGrab, detail NotifyAncestor, same_screen YES,
>     focus YES, state 0
> 
> FocusIn event, serial 34, synthetic NO, window 0x1c00001,
>     mode NotifyUngrab, detail NotifyAncestor
> 
> KeymapNotify event, serial 34, synthetic NO, window 0x0,
>     keys:  2   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
>            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
> 
> FocusOut event, serial 34, synthetic NO, window 0x1c00001,
>     mode NotifyGrab, detail NotifyAncestor
> 
> FocusIn event, serial 34, synthetic NO, window 0x1c00001,
>     mode NotifyUngrab, detail NotifyAncestor
> 
> KeymapNotify event, serial 34, synthetic NO, window 0x0,
>     keys:  2   0   0   0   0   0   0   0   4   0   0   0   0   0   0   0   
>            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
> 
> EnterNotify event, serial 34, synthetic NO, window 0x1c00001,
>     root 0x1a3, subw 0x0, time 23817793, (530,102), root:(1495,633),
>     mode NotifyUngrab, detail NotifyAncestor, same_screen YES,
>     focus YES, state 0
> 
> KeymapNotify event, serial 34, synthetic NO, window 0x0,
>     keys:  0   0   0   0   0   0   0   0   4   0   0   0   0   0   0   0   
>            0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
> 
> KeyRelease event, serial 34, synthetic NO, window 0x1c00001,
>     root 0x1a3, subw 0x0, time 23817872, (530,102), root:(1495,633),
>     state 0x0, keycode 66 (keysym 0xff1b, Escape), same_screen YES,
>     XLookupString gives 1 bytes: (1b) "
>     XFilterEvent returns: False
> 
> << The events below are me focussing back to the terminal that called xev >>
> 
> KeyPress event, serial 34, synthetic NO, window 0x1c00001,
>     root 0x1a3, subw 0x0, time 23819378, (530,102), root:(1495,633),
>     state 0x0, keycode 64 (keysym 0xffeb, Super_L), same_screen YES,
>     XLookupString gives 0 bytes: 
>     XmbLookupString gives 0 bytes: 
>     XFilterEvent returns: False
> 
> FocusOut event, serial 34, synthetic NO, window 0x1c00001,
>     mode NotifyGrab, detail NotifyAncestor
> 
> PropertyNotify event, serial 34, synthetic NO, window 0x1c00001,
>     atom 0x19a (_NET_FRAME_EXTENTS), time 23819452, state PropertyNewValue
> 
> ConfigureNotify event, serial 34, synthetic YES, window 0x1c00001,
>     event 0x1c00001, window 0x1c00001, (960,526), width 950, height 516,
>     border_width 5, above 0x0, override NO
> 
> PropertyNotify event, serial 34, synthetic NO, window 0x1c00001,
>     atom 0x198 (WM_STATE), time 23819453, state PropertyNewValue
> 
> FocusOut event, serial 34, synthetic NO, window 0x1c00001,
>     mode NotifyWhileGrabbed, detail NotifyNonlinear
> 
> FocusOut event, serial 34, synthetic NO, window 0x1c00001,
>     mode NotifyUngrab, detail NotifyPointer
> ```
> 
> The second event looks like a 'b' is being sent to xev, which made me think that I _may_ have read that xterm does not support send-event! However, testing with urxvt, kitty, and firefox, none of these show a 'b'. Here is a clip of me testing with firefox, which I include because you can actually see when I press the 'b' key that the focus to the text input kinda flashes off-on-off before rofi appears... not sure if that is relevant:
> out.mp4
> 
> What terminal are you testing with?

I am using "st" and there it works but not indeed on e.g. kitty, thanks for this 😅. Maybe XTest needs to be used instead? I will do some digging

**Comment by m-col** [comment]
> Maybe XTest needs to be used instead?

I think this is the likely solution. Clients would need to explicitly support send-event, adding a handler for those messages to their event loops. XTest on the other hand is 'invisible' to the clients and appears as if it is a real event, so will likely be the most seamless for its usage. Hopefully it also is on our implementation side.

**Comment by m-col** [comment]
You didn't actually need `simulate_keypress` in the end for the core code, right? I've been working on a wayland implementation and it is non-trivial I think. I'd be happy to drop that change if it is not needed for the swallowing feature. In case we do continue down this route, here is my current version in case it's of interest. It is completely untested and unfinished but you can see it gets a bit hacky:
```py
    def simulate_keypress(self, modifiers: List[str], key: str) -> None:
        """Simulates a keypress on the focused window."""
        if self.keyboards:
            #keysym = xkb.keysym_from_name(key, case_insensitive=True)
            #mods = wlrq.translate_masks(modifiers)
            event = wlr_ffi.new("struct wlr_event_keyboard_key *")
            event.time_msec = int(Timespec.get_monotonic_time().time * 1000)
            event.keycode
            event.update_state = True
            event.state = 1  # pressed
            self.keyboards[1].keyboard.key_event.emit(event)
            self.flush()
            event.state = 0  # released
            self.keyboards[1].keyboard.key_event.emit(event)
            self.flush()
```
The more "correct" alternative would be to create a virtual keyboard client and use that, which would work even if a keyboard wasn't currently attached to the seat (because it makes one). I haven't used client-side virtual keyboard code but I'm sure wlrctl would be handy for that.

**Comment by jwijenbergh** [comment]
> > Maybe XTest needs to be used instead?
> 
> I think this is the likely solution. Clients would need to explicitly support send-event, adding a handler for those messages to their event loops. XTest on the other hand is 'invisible' to the clients and appears as if it is a real event, so will likely be the most seamless for its usage. Hopefully it also is on our implementation side.

Yeah I'm implementing it atm, it's a bit more annoying but possible

**Comment by jwijenbergh** [comment]
> You didn't actually need `simulate_keypress` in the end for the core code, right? I've been working on a wayland implementation and it is non-trivial I think. I'd be happy to drop that change if it is not needed for the swallowing feature. In case we do continue down this route, here is my current version in case it's of interest. It is completely untested and unfinished but you can see it gets a bit hacky:
> 
> ```python
>     def simulate_keypress(self, modifiers: List[str], key: str) -> None:
>         """Simulates a keypress on the focused window."""
>         if self.keyboards:
>             #keysym = xkb.keysym_from_name(key, case_insensitive=True)
>             #mods = wlrq.translate_masks(modifiers)
>             event = wlr_ffi.new("struct wlr_event_keyboard_key *")
>             event.time_msec = int(Timespec.get_monotonic_time().time * 1000)
>             event.keycode
>             event.update_state = True
>             event.state = 1  # pressed
>             self.keyboards[1].keyboard.key_event.emit(event)
>             self.flush()
>             event.state = 0  # released
>             self.keyboards[1].keyboard.key_event.emit(event)
>             self.flush()
> ```
> 
> The more "correct" alternative would be to create a virtual keyboard client and use that, which would work even if a keyboard wasn't currently attached to the seat (because it makes one). I haven't used client-side virtual keyboard code but I'm sure wlrctl would be handy for that.

I indeed do not need it. It's just if the user wants to call it from the command line using `qtile cmd-obj -o cmd -f simulate_keypress --args "" "a"`. And indeed that looks not really trivial, but pretty similar to xtest actually.

**Comment by m-col** [comment]
I don't think we need to have a feature to simulate key presses, so if that command is not available (like currently) then that is fine. Keys can be simulated with external tools like `xdotool`, `wlrctl` etc.

**Comment by jwijenbergh** [comment]
> I don't think we need to have a feature to simulate key presses, so if that command is not available (like currently) then that is fine. Keys can be simulated with external tools like `xdotool`, `wlrctl` etc.

Agreed! However, I could easily add it for the x backend with xtest now. Can you test the latest commit? It is a bit WIP but it seems to work for me. I can't test extensively though right now. I will be able to work on it some more later. In the meantime let me know if there are any issues. It's maybe a bit hacky

**Comment by jwijenbergh** [comment]
Marked as draft up until we both find a solution that works for every situation. Hopefully the xtest approach I just pushed works :crossed_fingers:

File: libqtile/backend/x11/core.py
Comment (jwijenbergh) [inline libqtile/backend/x11/core.py:608]: Note to self, in `FakeInput` don't input all zeros but use the event settings. For DeviceId, we need to pass 0 though (last argument).

File: libqtile/backend/x11/core.py
Comment (jwijenbergh) [inline libqtile/backend/x11/core.py:628]: This is not the case when it is send from `simulate_keypress`, handle this or leave this to external tools such as xdotool? If we want to leave it to xdotool I would need to add a `simulated` argument again (with default `False`)

File: libqtile/backend/x11/core.py
Comment (jwijenbergh) [inline libqtile/backend/x11/core.py:631]: @m-col could it be the case that we're re-grabbing the key too early? Or does the `flush()` call already take care of it? I'm not sure if this is right what I'm doing.

File: libqtile/backend/x11/core.py
Comment (m-col) [reply libqtile/backend/x11/core.py:628]: Yeah, i think leaving that simulated keypress code the way it was is fine

File: libqtile/core/manager.py
Comment (jwijenbergh) [inline libqtile/core/manager.py:374]: I had to return a tuple with the key as I need to ungrab and grab it!

File: libqtile/backend/x11/core.py
Comment (jwijenbergh) [inline libqtile/backend/x11/core.py:627]: Note to self: test with KeyChords

File: libqtile/backend/x11/core.py
Comment (jwijenbergh) [reply libqtile/backend/x11/core.py]: I have done a middle ground, I have made simulate keypress closer to an x11 event so that we can in the future maybe support this. But for now simulated keypresses will only elicit qtile functions through the bindings, but will not actually forward it to a window. Xdotool is fine for this use case

**Comment by jwijenbergh** [comment]
Sorry for taking so long on this, was busy

Steps taken:
- Rebased with master
- Squashed commits to use xtest
- Simplified dummy events with simulate_keypress because we just want to advise users to use xdotool/ydotool
- Tested manually with keychord and keys, seems to work fine for me

**Comment by elParaguayo** [comment]
You probably need to add `xcffib.xtest` to `MOCK_MODULES` in `docs/conf.py`.

**Comment by jwijenbergh** [comment]
> You probably need to add `xcffib.xtest` to `MOCK_MODULES` in `docs/conf.py`.

Thanks! Was wondering why they were failing

File: libqtile/core/manager.py
Comment (m-col) [inline libqtile/core/manager.py:410]: We no longer need Optional, Tuple or Union:
```suggestion
    ) -> tuple[Key | KeyChord | None, bool]:
```

**Review by m-col — COMMENTED** [review]
Excellent work as always. For the pep8 complaint feel free to `# noqa: N802` it or change the name of the method, up to you. Then LGTM!

File: libqtile/core/manager.py
Comment (jwijenbergh) [reply libqtile/core/manager.py:410]: Ah yes thanks, that is really neat

**Comment by jwijenbergh** [comment]
Fixes pushed, I have named it `fake_KeyPress` again with the noqa command to be in line with `handle_KeyPress`

**Comment by m-col** [comment]
Do you mind if I push some commits? I've been playing around with some of the event simulation for the tests

**Comment by jwijenbergh** [comment]
> Do you mind if I push some commits? I've been playing around with some of the event simulation for the tests

Go ahead, that would be great! Do I need to enable that somewhere?

**Comment by jwijenbergh** [comment]
In the meantime let me figure out if it's maybe possible to do pointer grabs async as well and use xtest too. I think maybe grabbing things sync is not the best idea

**Comment by jwijenbergh** [comment]
Steps taken:

- Rebased with master
- Fixed changelog entry (it was in the wrong release lol)
- Removed ability to swallow for mouse bindings as grabbing sync is not really nice IMO and I did not get it working with xtest. On top of that I think that not swallowing mouse bindings is less useful. So for now it's only Key and KeyChord.

**Comment by m-col** [comment]
> Go ahead, that would be great! Do I need to enable that somewhere?

Yes, I think you will have something in the sidebar that will enable this?

**Comment by m-col** [comment]
I didn't want to push to this directly as I'm unsure about all the changes I made. Two commits here: https://github.com/m-col/qtile/commits/3113

The first commit exposes commands for `warp_pointer` and `simulate_click` which makes it easier to perform these behaviours in tests in a backend-agnostic way.

The second commit doesn't work and doesn't fix anything, it was a bit of a work in progress before I got distracted. There were two things I had in mind with the tests. Less important was using `test_data` to count swallows to give a bit more of a quantitative assessment of what's going on in the tests. More important was removing the x11-specific imports in `test_swallow.py`. Qtile should be able to run under Wayland without xcffib installed, and the same for running the tests with the wayland backend only (`--backend=wayland`). Currently xcffib is required unconditionally because xcbq is imported in `confreader.py` but I have some work in the pipeline to make xcffib unnecessary for systems that are wayland only. Either way we should merge this asap, so I'm happy for us to either move the tests into `test/backend/x11/` or remove them completely in the mean time (probably keeping them for x11 only is preferable though!).

**Comment by jwijenbergh** [comment]
> > Go ahead, that would be great! Do I need to enable that somewhere?
> 
> Yes, I think you will have something in the sidebar that will enable this?

Ah yes this should be on

**Comment by jwijenbergh** [comment]
Steps taken:
- Implemented some changes by @m-col to cleanup code and improve tests
- Rebased with master

File: libqtile/core/manager.py
Comment (m-col) [inline libqtile/core/manager.py:412]: The bool here specifies whether the binding was handled, and if it wasn't then it couldn't be swallowed, right?

What I'm thinking is that the `tuple[Key | KeyChord | None, bool]` return type implies that `(None, True)` is a valid return value, but from what I can tell it wouldn't be valid. Could it be `tuple[Key | KeyChord, bool] | None` instead? This would also mean the change in `libqtile/backend/wayland/inputs.py` doesn't need the `[1]` index, making it a bit more clear when just reading the code in that file.

**Review by m-col — APPROVED** [review]
Just one comment but other than that, LGTM :)

File: libqtile/core/manager.py
Comment (jwijenbergh) [reply libqtile/core/manager.py:412]: But if we return ```(key, False)``` then before it would return ```False```, whereas with the change it would return ```True``` if I don't index by 1. This would only work for the ```None, False``` case where we can now just return ```None```.

 I could also make it ```Key | KeyChord | None``` but then I would have to set key to None conditionally on the ```key.swallow``` parameter which is weird when reading ```process_key_event```. What do you think?

**Comment by jwijenbergh** [comment]
Rebased with master.

File: libqtile/core/manager.py
Comment (m-col) [reply libqtile/core/manager.py:412]: OK, nvm my suggestion actually. I think it's fine as is :)

**Comment by m-col** [comment]
Happy to merge this after rebasing (assuming that fixes the CI) :)

File: libqtile/core/manager.py
Comment (jwijenbergh) [reply libqtile/core/manager.py:419]: Great, thanks for thinking with me :+1:

**Pull request merged**
